### DISCUSSION TITLE
when a form has both `action` and `onSubmit`, in which order are they called?

### DISCUSSION THREAD

**Discussion by stefanprobst** [vercel/next.js#63435] [App Router] [answered]
i am wondering if there is a guaranteed order in which callbacks on the `action` and `onSubmit` props on a `<form>` are called?

**Comment by ModulesSoft** [comment, marked as answer]
The `onSubmit` callback is called first, allowing you to handle the form submission within your React component. 
It is called before the form is actually submitted.
Then, if the form's `action` attribute is specified, the browser navigates to the specified URL.

**Comment by dw820** [comment]
Hi @ModulesSoft I have a follow up question, so if we used client side validation with react-hook-form on onSubmit, action will still be triggered even if the client side validation fail? Thanks!

**Reply by ModulesSoft** [reply]
I'm not familiar with React Hook Form. 
Yet, I would say if client-side validation fails during the onSubmit event, the action should not be triggered.
